src/wallet/PegOut.cpp

```diff
@@ -0,0 +1,85 @@
+#include "PegOut.h"
+#include "CoinSelection.h"
+#include "WalletUtil.h"
+
+#include <mw/crypto/Blinds.h>
+#include <mw/exceptions/InsufficientFundsException.h>
+#include <mw/wallet/Wallet.h>
+#include <mw/wallet/KernelFactory.h>
+#include <mw/wallet/OutputFactory.h>
+#include <mw/wallet/TxFactory.h>
+
+// TODO: Sort inputs, outputs, kernels, and owner_sigs
+mw::Transaction::CPtr PegOut::CreatePegOutTx(
+    const uint64_t amount,
+    const uint64_t fee_base,
+    const Bech32Address& address) const
+{
+    // Select coins
+    std::vector<libmw::Coin> input_coins = CoinSelection::SelectCoins(m_wallet.GetInterface(), amount, fee_base);
+    std::vector<Input> inputs = WalletUtil::SignInputs(input_coins);
+
+    // Calculate fee
+    const uint64_t fee = WalletUtil::CalculateFee(fee_base, input_coins.size(), 1, 2);
+    if (WalletUtil::TotalAmount(input_coins) <= (amount + fee)) {
+        ThrowInsufficientFunds("Not enough funds");
+    }
+
+    // Create change output.
+    // We randomly generate the sender_key and output_blind.
+    // Receiver key is generate by OutputFactory using the wallet's stealth address.
+    const uint64_t change_amount = WalletUtil::TotalAmount(input_coins) - (amount + fee);
+    SecretKey change_key = m_wallet.NewKey();
+    BlindingFactor change_blind = Random().CSPRNG<32>();
+    Output change_output = OutputFactory::Create(
+        EOutputFeatures::DEFAULT_OUTPUT,
+        change_blind,
+        change_key,
+        m_wallet.GetStealthAddress(),
+        change_amount
+    );
+
+    // Total kernel offset is split between raw kernel_offset and the kernel's blinding factor.
+    // sum(output.blind) - sum(input.blind) = kernel_offset + sum(kernel.blind)
+    BlindingFactor kernel_offset = Random::CSPRNG<32>();
+    std::vector<BlindingFactor> input_blinds = WalletUtil::GetBlindingFactors(input_coins);
+    BlindingFactor kernel_blind = Blinds()
+        .Add(change_blind)
+        .Sub(input_blinds)
+        .Sub(kernel_offset)
+        .Total();
+    Kernel kernel = KernelFactory::CreatePegOutKernel(kernel_blind, amount, fee, address);
+
+    // TODO: Only necessary when no change?
```

Comment:
Just wondering, why owner sig might not be required if there is change? Should it be 'only necessary when there is change'?

Reply:
There's a very long answer to that which would best be done as a blog post. I will prioritize that.